validator/execution_run.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package validator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+type ExecutionRun interface {
+	GetStepAt(uint64) MachineStep
+	GetLastStep() MachineStep
+	PrepareRange(uint64, uint64)
+	Close()
+}
+
+type MachineStep interface {
+	ReadyMarker
+	Hash() (common.Hash, error)
```

Comment:
I made all long-delay rpcs a Promise (ReadyMarker at that stage).
I think these MachineSteps are indeed always just used by Await - but even so, this property makes sense to me. It also enforces some interesting properties that I like, separating the client context which is used for the rpc from the context passed to GetHashAt.
Later, I think it could be used to add connection management (e.g. retries).